### DISCUSSION TITLE
Any way to specify metadata within a razor page rather than using yaml/json?

### DISCUSSION THREAD

**Discussion by Khitiara** [statiqdev/Discussions#132] [Q&A] [answered]
I'd like to specify metadata either through a page model or through attributes on the razor page rather than using a seperate yaml/json frontmatter section or sidecar file, is this possible?

**Comment by daveaglick** [comment, marked as answer]
Not really. The Razor engine exists at a level "higher" than the Statiq document model. Document metadata can flow up through the document to be exposed in the Razor template via the `Document` property (and in some cases `Model` - though `Document` is preferred). But we can't flow metadata back down to the document through the Razor template.

Part of the reasoning here is that it would be confusing. Razor only executes near the very end of the Statiq process in the [post-process phase](https://www.statiq.dev/framework/pipelines/#phases) so even if we did allow metadata to be added to a document from within Razor, it wouldn't really be available to much else after the Razor engine in done processing.

On the other hand, there's no reason you can't set common properties or other values like you normally would in Razor. For example, adding this to the top of your layout would work fine:

```
@{
    string MyValue = "foo";
}
```

And `MyValue` would be available as a variable in the rest of the Razor page. You also have access to the `ViewBag` just like you do in a normal Razor environment so that can be used to pass around variables as well.

What in particular are you trying to accomplish? Maybe that will help me understand what the use case here is.

**Reply by Khitiara** [reply]
I was mainly thinking that the yaml front matter feels out of place and potentially confusing in an otherwise razor and code-based project; at least personally I find the front matter acts in a kinda magic way in much the same way assembly scanning often does, obscuring the purpose and also interfering a lot with preview tools so I was wondering if there was an attribute or otherwise code-based way to establish the metadata

**Reply by daveaglick** [reply]
Ah - I understand. Yeah, because of the way Statiq is template-agnostic, using a particular templating engine to define metadata would be challenging and likely produce very unexpected results due to some templating engines like Razor being evaluated well after other processing that uses metadata is done (I.e. "why isn't the title metadata I defined on this page being used for archives"). I'll chew on this a bit, but I'm doubtful there's a way around using front-matter or sidecar files.

What about front-matter formatted as special code comments? That might be doable since parsing the front-matter would only require reading the Razor file, not compiling/evaluating it. Something like:

`myfile.cshtml`:
```
@*---
Title: Foo
Date: 1/1/2022
Color: Blue
---*@

...the rest of the Razor file
```

(I'd need to experiment with syntax, but you get the idea).

**Reply by Khitiara** [reply]
special comments makes sense to me

**Reply by daveaglick** [reply]
Awesome - comment style front-matter is already supported for [scripts](https://www.statiq.dev/web/content-and-data/scripts#front-matter) so hopefully adding something similar for Razor won't be a big lift. I'll see what I can do for the next release.

**Reply by daveaglick** [reply]
Added https://github.com/statiqdev/Statiq.Framework/issues/231 to track this work
